### PR TITLE
Add initial implementation

### PR DESCRIPTION
**Pull request by joaonice**
This adds the initial implementation of `objection-unique`.

This plugin adds a unique validation for `objection` models.

### CODE REVIEW

File: .eslintrc.yml [hunk 1]
```diff
@@ -0,0 +1,12 @@
+extends: seegno
+
+env:
+  jasmine: true
+  jest: true
+
+plugins: [jest]
```
Comment (abelsoares) [inline .eslintrc.yml]:
```
plugins: 
  - jest
```

File: README.md [hunk 2]
```diff
@@ -1,2 +1,122 @@
...
+
+## Tests
+
+Run the tests from the root directory:
+
+```sh
+npm test
+```
+
+You can find Jest documentation [here](https://facebook.github.io/jest/docs/getting-started.html).
```
Comment (abelsoares) [inline README.md]: I would remove this.

File: index.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+/**
+ * Module dependencies.
+ */
+
+const isEmpty = require('lodash.isempty');
+const isUndefined = require('lodash.isundefined');
+const isNull = require('lodash.isnull');
```
Comment (abelsoares) [inline index.js]: Sort.

File: package.json [hunk 4]
```diff
@@ -0,0 +1,86 @@
+{
+  "name": "objection-unique",
+  "version": "2.0.0",
```
Comment (abelsoares) [inline package.json]: Is this the right version?

File: test/index.test.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+
+/**
+ * Test `FoobarController`.
+ */
+
+describe('FoobarController', () => {
+  beforeEach(clearDatabase);
+
+  describe('$beforeInsert', () => {
+    it('should throw an error if there is no fields or identifiers options.`', async () => {
```
Comment (abelsoares) [inline test/index.test.js]: Extra backtick at the end. Remove all in this file.

File: test/index.test.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+    });
+
+    it('should insert the given data.', async () => {
+      const TestModel = modelFactory({
+        fields: ['bar', 'foo']
+      });
+
+      const data = { bar: 'bar', biz: 'biz', foo: 'foo' };
+      const { id } = await TestModel.query().insert(data);
+      const test = await TestModel.query().findById(id);
```
Comment (abelsoares) [inline test/index.test.js]: WDYT about call this `result`?

File: test/index.test.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+
+      expect(test).toEqual({ ...data, biz: null, id });
+    });
+
+    it('should insert the given data.', async () => {
+      const TestModel = modelFactory({
+        fields: ['bar', 'foo']
+      });
+
+      const data = { bar: 'bar', biz: 'biz', foo: 'foo' };
```
Comment (abelsoares) [inline test/index.test.js]: WDYT about passing this directly as argument. Also, in the other tests like this one.

File: README.md [hunk 8]
```diff
@@ -1,2 +1,118 @@
...
+
+## Contributing & Development
+
+#### Contributing
```
Comment (nunorafaelrocha) [inline README.md]: Use `###`

File: index.js [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+        }
+
+        if (isEmpty(queryOptions.old)) {
+          throw new Error('Unique validation at update only works with queries started with $query.');
+        }
+
+        return this.queryResolver(parent, true, queryOptions);
+      }
+
+      queryResolver(parent, update = false, queryOptions = {}) {
```
Comment (nunorafaelrocha) [inline index.js]: Add missing dock block.

File: index.js [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+       */
+
+      getQuery(update, queryOptions) {
+        return options.fields.reduce((queries, field) => {
+          if (isNil(this[field])) {
+            return queries;
+          }
+
+          const knex = Model.knex();
+
```
Comment (nunorafaelrocha) [inline index.js]: Remove new line.

File: package.json [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+  "name": "objection-unique",
+  "version": "0.0.0",
+  "description": "objection-unique",
+  "keywords": [
+    "entity",
+    "objection",
+    "objectionjs",
+    "plugin",
+    "plugins",
+    "unique"
```
Comment (nunorafaelrocha) [inline package.json]: Add validation related keywords.

File: test/index.test.js [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+      }
+    });
+
+    it('should update the entity ignoring the unique validation if the values are from the same entity that are begin updated.', async () => {
+      const TestModel = modelFactory({
+        fields: ['bar', 'foo']
+      });
+
+      let result = await TestModel.query().insertAndFetch({ bar: 'biz', foo: 'buz' });
+
```
Comment (nunorafaelrocha) [inline test/index.test.js]: Remove new line.

File: test/index.test.js [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+      expect(result).toEqual({ ...result, biz: 'foo' });
+    });
+
+    it('should update the entity ignoring null values validation.`', async () => {
+      const TestModel = modelFactory({
+        fields: ['bar', 'foo']
+      });
+
+      let result = await TestModel.query().insert({ bar: 'bar', foo: null });
+
```
Comment (nunorafaelrocha) [inline test/index.test.js]: Remove new line.

File: test/index.test.js [hunk 14]
```diff
@@ -0,0 +1,172 @@
...
+      expect(result).toEqual({ bar: 'biz', biz: 'waldo', foo: null, id: result.id });
+    });
+
+    it('should update the entity.`', async () => {
+      const TestModel = modelFactory({
+        fields: ['bar', 'foo']
+      });
+
+      let result = await TestModel.query().insertAndFetch({ bar: 'bar', foo: 'foo' });
+
```
Comment (nunorafaelrocha) [inline test/index.test.js]: Remove new line.

File: test/utils/clear-database.js [hunk 15]
```diff
@@ -0,0 +1,16 @@
+
+/**
+ * Module dependencies.
+ */
+
+import modelFactory from './model-factory';
+
+/**
+ * Export `setupDatabase`.
```
Comment (nunorafaelrocha) [inline test/utils/clear-database.js]: Hmmm... clearDatabase?

File: test/utils/knex.js [hunk 16]
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * Module dependencies.
+ */
+
+import knex from 'knex';
+
+// Connection configuration.
```
Comment (nunorafaelrocha) [inline test/utils/knex.js]: Knex configuration? Database configuration?

File: README.md [hunk 17]
```diff
@@ -1,2 +1,118 @@
...
+This plugin adds a unique validation for [Objection.js](https://github.com/Vincit/objection.js/) models.
+
+**NOTE:** Unique validation at update only works with `$query` methods.
+
+## Installation
+
+### NPM
+
+```sh
+npm i objection-unique
```
Comment (nunorafaelrocha) [inline README.md]: `npm install objection-unique --save`?

File: README.md (hunk 2 above)
Comment (nunorafaelrocha) [inline README.md]: Me too ;)

File: index.js [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+          const query = knex(this.constructor.tableName)
+            .select()
+            .where(field, this[field])
+            .limit(1);
+
+          if (update) {
+            options.identifiers.forEach(identifier => {
+              query.andWhereNot(identifier, queryOptions.old[identifier]);
+            });
+          }
```
Comment (abelsoares) [inline index.js]: Maybe:
```
if (update) {
  options.identifiers.forEach(identifier =>
    query.andWhereNot(identifier, queryOptions.old[identifier])
  );
}
```

**Pull request merged**